File: csharp/1386.cinema-seat-allocation.cs
/*
 * @lc app=leetcode id=1386 lang=csharp
 *
 * [1386] Cinema Seat Allocation
 */

using System.Collections.Generic;
using System.Linq;

// @lc code=start
public class Solution {
    public int MaxNumberOfFamilies(int n, int[][] reservedSeats) {
        var taken = new Dictionary<int, int>();
        foreach (var t in reservedSeats) {
            if (!taken.TryGetValue(t[0], out var v)) {
                v = 0;
            }
            taken[t[0]] = v | (1 << t[1]);
        }
        int ret = 2 * (n - taken.Count);
        foreach (var p in taken) {
            int a = 0, t = p.Value;
            if ((t & 60) == 0) {
                ++a;
            }
            if ((t & 960) == 0) {
                ++a;
            }
            if (a == 0 && (t & 240) == 0) {
                ++a;
            }
            ret += a;
        }
        return ret;
    }
}
// @lc code=end

File: java/1386.cinema-seat-allocation.java
/*
 * @lc app=leetcode id=1386 lang=java
 *
 * [1386] Cinema Seat Allocation
 */

import java.util.HashMap;
import java.util.Map;

// @lc code=start
class Solution {
    public int maxNumberOfFamilies(int n, int[][] reservedSeats) {
        Map<Integer, Integer> rows = new HashMap<>();
        for (int[] ss : reservedSeats) {
            rows.put(ss[0] - 1, rows.getOrDefault(ss[0] - 1, 0) | (1 << (ss[1] - 1)));
        }
        int ret = 2 * (n - rows.size());
        for (int row : rows.values()) {
            if ((row & 0b11110) == 0) {
                ++ret;
                if ((row & 0b111100000) == 0) {
                    ++ret;
                }
            } else if ((row & 0b1111000) == 0) {
                ++ret;
            } else if ((row & 0b111100000) == 0) {
                ++ret;
            }
        }
        return ret;
    }
}
// @lc code=end

